# STEP3: Perform some simple data analysis

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed steps 1 and 2, then skip to the second cell.

In [ ]:
# !PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila
# !PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila -f Data/pagila-schema.sql
# !PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila -f Data/pagila-data.sql

In [1]:
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila'
DB_USER = 'Moaze002'
DB_PASSWORD = 'admin'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://Moaze002:admin@127.0.0.1:5432/pagila


### 3NF - Entity Relationship Diagram

<img src="assets/pagila-3nf.png" width="50%"/>

## 3.1 Insight 1:   Top Grossing Movies 
- Payments amounts are in table `payment`
- Movies are in table `film`
- They are not directly linked, `payment` refers to a `rental`, `rental` refers to an `inventory` item and `inventory` item refers to a `film`
- `payment` &rarr; `rental` &rarr; `inventory` &rarr; `film`

### 3.1.1 Films

In [2]:
%%sql
select film_id, title, release_year, rental_rate, rating  from film limit 5;


 * postgresql://Moaze002:***@127.0.0.1:5432/pagila
5 rows affected.


[(1, 'ACADEMY DINOSAUR', 2012, Decimal('0.99'), 'PG'),
 (2, 'ACE GOLDFINGER', 2023, Decimal('4.99'), 'G'),
 (3, 'ADAPTATION HOLES', 2017, Decimal('2.99'), 'NC-17'),
 (4, 'AFFAIR PREJUDICE', 2023, Decimal('2.99'), 'G'),
 (5, 'AFRICAN EGG', 2019, Decimal('2.99'), 'G')]

### 3.1.2 Payments

In [3]:
%%sql
select * from payment limit 5;

 * postgresql://Moaze002:***@127.0.0.1:5432/pagila
5 rows affected.


[(16051, 269, 1, 98, Decimal('0.99'), datetime.datetime(2022, 1, 29, 2, 58, 52, 222594, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600)))),
 (16065, 274, 1, 147, Decimal('2.99'), datetime.datetime(2022, 1, 25, 13, 14, 16, 895377, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600)))),
 (16109, 297, 2, 143, Decimal('0.99'), datetime.datetime(2022, 1, 28, 1, 49, 49, 128218, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600)))),
 (16195, 344, 2, 157, Decimal('2.99'), datetime.datetime(2022, 1, 31, 6, 58, 51, 176578, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600)))),
 (16202, 348, 2, 821, Decimal('0.99'), datetime.datetime(2022, 1, 26, 17, 52, 41, 359433, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))))]

### 3.1.3 Inventory

In [8]:
%%sql
select * from inventory limit 5;

 * postgresql://Moaze002:***@127.0.0.1:5432/pagila
5 rows affected.


[(1, 1, 1, datetime.datetime(2022, 2, 15, 11, 9, 17, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600)))),
 (2, 1, 1, datetime.datetime(2022, 2, 15, 11, 9, 17, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600)))),
 (3, 1, 1, datetime.datetime(2022, 2, 15, 11, 9, 17, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600)))),
 (4, 1, 1, datetime.datetime(2022, 2, 15, 11, 9, 17, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600)))),
 (5, 1, 2, datetime.datetime(2022, 2, 15, 11, 9, 17, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))))]

### 3.1.4 Get the movie of every payment

In [9]:
%%sql
SELECT f.title, p.amount, p.payment_date, p.customer_id                                            
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id )
JOIN film f ON ( i.film_id = f.film_id)
limit 5;

 * postgresql://Moaze002:***@127.0.0.1:5432/pagila
5 rows affected.


[('PACKER MADIGAN', Decimal('0.99'), datetime.datetime(2022, 1, 29, 2, 58, 52, 222594, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))), 269),
 ('CHARIOTS CONSPIRACY', Decimal('2.99'), datetime.datetime(2022, 1, 25, 13, 14, 16, 895377, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))), 274),
 ('CRAFT OUTFIELD', Decimal('0.99'), datetime.datetime(2022, 1, 28, 1, 49, 49, 128218, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))), 297),
 ('CRUSADE HONEY', Decimal('2.99'), datetime.datetime(2022, 1, 31, 6, 58, 51, 176578, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))), 344),
 ('ROMAN PUNK', Decimal('0.99'), datetime.datetime(2022, 1, 26, 17, 52, 41, 359433, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))), 348)]

### 3.1.5 sum movie rental revenue
Write a query that displays the amount of revenue from each title. Limit the results to the top 10 grossing titles. Your results should match the table below.

In [46]:
%%sql
WITH revenue as 
    (SELECT f.title, sum(p.amount) AS revenue                                      
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id )
JOIN film f ON ( i.film_id = f.film_id)
    GROUP BY f.title)

SELECT * FROM revenue
ORDER BY revenue DESC
limit 10;

 * postgresql://Moaze002:***@127.0.0.1:5432/pagila
10 rows affected.


[('TELEGRAPH VOYAGE', Decimal('231.73')),
 ('WIFE TURN', Decimal('223.69')),
 ('ZORRO ARK', Decimal('214.69')),
 ('GOODFELLAS SALUTE', Decimal('209.69')),
 ('SATURDAY LAMBS', Decimal('204.72')),
 ('TITANS JERK', Decimal('201.71')),
 ('TORQUE BOUND', Decimal('198.72')),
 ('HARRY IDAHO', Decimal('195.70')),
 ('INNOCENT USUAL', Decimal('191.74')),
 ('HUSTLER PARTY', Decimal('190.78'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>title</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>TELEGRAPH VOYAGE</td>
        <td>231.73</td>
    </tr>
    <tr>
        <td>WIFE TURN</td>
        <td>223.69</td>
    </tr>
    <tr>
        <td>ZORRO ARK</td>
        <td>214.69</td>
    </tr>
    <tr>
        <td>GOODFELLAS SALUTE</td>
        <td>209.69</td>
    </tr>
    <tr>
        <td>SATURDAY LAMBS</td>
        <td>204.72</td>
    </tr>
    <tr>
        <td>TITANS JERK</td>
        <td>201.71</td>
    </tr>
    <tr>
        <td>TORQUE BOUND</td>
        <td>198.72</td>
    </tr>
    <tr>
        <td>HARRY IDAHO</td>
        <td>195.70</td>
    </tr>
    <tr>
        <td>INNOCENT USUAL</td>
        <td>191.74</td>
    </tr>
    <tr>
        <td>HUSTLER PARTY</td>
        <td>190.78</td>
    </tr>
</tbody></table></div>

## 3.2 Insight 2:   Top grossing cities 
- Payments amounts are in table `payment`
- Cities are in table `cities`
- `payment` &rarr; `customer` &rarr; `address` &rarr; `city`

### 3.2.1 Get the city of each payment

In [4]:
%%sql
SELECT p.customer_id, p.rental_id, p.amount, ci.city                            
FROM payment p
JOIN customer c  ON ( p.customer_id = c.customer_id )
JOIN address a ON ( c.address_id = a.address_id )
JOIN city ci ON ( a.city_id = ci.city_id )
order by p.payment_date
limit 10;

 * postgresql://Moaze002:***@127.0.0.1:5432/pagila
10 rows affected.


[(427, 8182, Decimal('3.99'), 'Guaruj'),
 (180, 2798, Decimal('2.99'), 'Pereira'),
 (426, 10172, Decimal('10.99'), 'Purnea (Purnia)'),
 (295, 2054, Decimal('2.99'), 'Kolpino'),
 (172, 1507, Decimal('0.99'), 'Batman'),
 (277, 6487, Decimal('5.99'), 'Ogbomosho'),
 (2, 8705, Decimal('5.99'), 'San Bernardino'),
 (108, 4082, Decimal('2.99'), 'Huixquilucan'),
 (141, 13470, Decimal('4.99'), 'Fukuyama'),
 (43, 8376, Decimal('4.99'), 'Faaa')]

### 3.2.2 Top grossing cities
Write a query that returns the total amount of revenue by city as measured by the `amount` variable in the `payment` table. Limit the results to the top 10 cities. Your result should match the table below.

In [9]:
%%sql

WITH customer_rev AS (
    SELECT sum(p.amount) AS revenue, ci.city                            
FROM payment p
JOIN customer c  ON p.customer_id = c.customer_id
JOIN address a   ON c.address_id = a.address_id
JOIN city ci     ON a.city_id = ci.city_id
GROUP BY  ci.city
)

SELECT * FROM customer_rev
ORDER BY revenue DESC   
limit 10;



 * postgresql://Moaze002:***@127.0.0.1:5432/pagila
10 rows affected.


[(Decimal('221.55'), 'Cape Coral'),
 (Decimal('216.54'), 'Saint-Denis'),
 (Decimal('198.50'), 'Aurora'),
 (Decimal('195.58'), 'Molodetno'),
 (Decimal('194.61'), 'Santa Brbara dOeste'),
 (Decimal('194.61'), 'Apeldoorn'),
 (Decimal('186.62'), 'Qomsheh'),
 (Decimal('180.52'), 'London'),
 (Decimal('177.60'), 'Ourense (Orense)'),
 (Decimal('175.61'), 'Bijapur')]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>Cape Coral</td>
        <td>221.55</td>
    </tr>
    <tr>
        <td>Saint-Denis</td>
        <td>216.54</td>
    </tr>
    <tr>
        <td>Aurora</td>
        <td>198.50</td>
    </tr>
    <tr>
        <td>Molodetno</td>
        <td>195.58</td>
    </tr>
    <tr>
        <td>Apeldoorn</td>
        <td>194.61</td>
    </tr>
    <tr>
        <td>Santa Brbara dOeste</td>
        <td>194.61</td>
    </tr>
    <tr>
        <td>Qomsheh</td>
        <td>186.62</td>
    </tr>
    <tr>
        <td>London</td>
        <td>180.52</td>
    </tr>
    <tr>
        <td>Ourense (Orense)</td>
        <td>177.60</td>
    </tr>
    <tr>
        <td>Bijapur</td>
        <td>175.61</td>
    </tr>
</tbody></table></div>

## 3.3 Insight 3 : Revenue of a movie by customer city and by month 

### 3.3.1 Total revenue by month

In [10]:
%%sql
SELECT sum(p.amount) as revenue, EXTRACT(month FROM p.payment_date) as month
from payment p
group by month
order by revenue desc
limit 10;

 * postgresql://Moaze002:***@127.0.0.1:5432/pagila
7 rows affected.


[(Decimal('11413.88'), Decimal('3')),
 (Decimal('11371.24'), Decimal('5')),
 (Decimal('10897.49'), Decimal('6')),
 (Decimal('10738.56'), Decimal('4')),
 (Decimal('10151.99'), Decimal('2')),
 (Decimal('9751.56'), Decimal('7')),
 (Decimal('3091.79'), Decimal('1'))]

### 3.3.2 Each movie by customer city and by month (data cube)

In [11]:
%%sql
SELECT f.title, p.amount, p.customer_id, ci.city, p.payment_date,EXTRACT(month FROM p.payment_date) as month
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id )
JOIN film f ON ( i.film_id = f.film_id)
JOIN customer c  ON ( p.customer_id = c.customer_id )
JOIN address a ON ( c.address_id = a.address_id )
JOIN city ci ON ( a.city_id = ci.city_id )
order by p.payment_date
limit 10;

 * postgresql://Moaze002:***@127.0.0.1:5432/pagila
10 rows affected.


[('TOMORROW HUSTLER', Decimal('3.99'), 427, 'Guaruj', datetime.datetime(2022, 1, 23, 14, 3, 52, 212496, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))), Decimal('1')),
 ('EXPRESS LONELY', Decimal('2.99'), 180, 'Pereira', datetime.datetime(2022, 1, 23, 14, 24, 17, 906429, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))), Decimal('1')),
 ('MAIDEN HOME', Decimal('10.99'), 426, 'Purnea (Purnia)', datetime.datetime(2022, 1, 23, 14, 42, 35, 952907, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))), Decimal('1')),
 ('GANDHI KWAI', Decimal('2.99'), 295, 'Kolpino', datetime.datetime(2022, 1, 23, 14, 43, 42, 505434, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))), Decimal('1')),
 ('MAJESTIC FLOATS', Decimal('0.99'), 172, 'Batman', datetime.datetime(2022, 1, 23, 14, 57, 4, 87741, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))), Decimal('1')),
 ('SWEDEN SHINING', Decimal('5.99'), 277, 'Ogbomosho', datetime.datetime(2022, 1, 23, 15, 5, 24, 118128, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))), Decimal('1')),
 ('TELEGRAPH VOYAGE', Decimal('5.99'), 2, 'San Bernardino', datetime.datetime(2022, 1, 23, 15, 26, 35, 170413, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))), Decimal('1')),
 ('OPERATION OPERATION', Decimal('2.99'), 108, 'Huixquilucan', datetime.datetime(2022, 1, 23, 15, 44, 27, 976362, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))), Decimal('1')),
 ('PINOCCHIO SIMON', Decimal('4.99'), 141, 'Fukuyama', datetime.datetime(2022, 1, 23, 16, 6, 30, 830136, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))), Decimal('1')),
 ('SHAWSHANK BUBBLE', Decimal('4.99'), 43, 'Faaa', datetime.datetime(2022, 1, 23, 16, 16, 12, 762605, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))), Decimal('1'))]

### 3.3.3 Sum of revenue of each movie by customer city and by month

Write a query that returns the total amount of revenue for each movie by customer city and by month. Limit the results to the top 10 movies. Your result should match the table below.

In [17]:
%%sql
WITH customer_film_revenue AS (
    SELECT f.title, sum(p.amount) as revenue, ci.city ,EXTRACT(month FROM p.payment_date) as month
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id )
JOIN film f ON ( i.film_id = f.film_id)
JOIN customer c  ON ( p.customer_id = c.customer_id )
JOIN address a ON ( c.address_id = a.address_id )
JOIN city ci ON ( a.city_id = ci.city_id )
GROUP BY f.title, ci.city, month
)

SELECT * FROM customer_film_revenue
ORDER BY revenue
limit 10;

 * postgresql://Moaze002:***@127.0.0.1:5432/pagila
10 rows affected.


[('CYCLONE FAMILY', Decimal('0.00'), 'Phnom Penh', Decimal('5')),
 ('HALF OUTFIELD', Decimal('0.00'), 'Yaound', Decimal('7')),
 ('CHASING FIGHT', Decimal('0.00'), 'Bern', Decimal('7')),
 ('CURTAIN VIDEOTAPE', Decimal('0.00'), 'Balurghat', Decimal('4')),
 ('DEER VIRGINIAN', Decimal('0.00'), 'Allappuzha (Alleppey)', Decimal('6')),
 ('FRIDA SLIPPER', Decimal('0.00'), 'Benin City', Decimal('5')),
 ('BLADE POLISH', Decimal('0.00'), 'Merlo', Decimal('2')),
 ('CHAMBER ITALIAN', Decimal('0.00'), 'Jaipur', Decimal('5')),
 ('CLEOPATRA DEVIL', Decimal('0.00'), 'Rockford', Decimal('1')),
 ('HOLES BRANNIGAN', Decimal('0.00'), 'Kakamigahara', Decimal('1'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>title</th>
        <th>city</th>
        <th>month</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>SHOW LORD</td>
        <td>Mannheim</td>
        <td>1.0</td>
        <td>11.99</td>
    </tr>
    <tr>
        <td>AMERICAN CIRCUS</td>
        <td>Callao</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>CASUALTIES ENCINO</td>
        <td>Warren</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>TELEGRAPH VOYAGE</td>
        <td>Naala-Porto</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>KISSING DOLLS</td>
        <td>Toulon</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>MILLION ACE</td>
        <td>Bergamo</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>TITANS JERK</td>
        <td>Kimberley</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>DARKO DORADO</td>
        <td>Bhilwara</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>SUNRISE LEAGUE</td>
        <td>Nagareyama</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>MILLION ACE</td>
        <td>Gaziantep</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
</tbody></table></div>